File: src/main/java/com/efo/fjospidie/Engines/Snort/SnortEngine.java
package com.efo.fjospidie.engines.snort;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.ProcessBuilder.Redirect;
import java.util.ArrayList;
import java.util.List;

import com.efo.fjospidie.FjoSpidie;
import com.efo.fjospidie.LOG;
import com.efo.fjospidie.Report;

public class SnortEngine extends Thread {

	List<String> connections;
	private List<SnortAlert> alerts = new ArrayList<SnortAlert>();

	public SnortEngine(List<String> connections) {
		this.connections = connections;
	}

	public void run() {

		try {
			Process analisysProgess;
			ProcessBuilder pb;
			String EXTERNAL_NET = "[";
			for (String connection : connections) {
				if (connection != null)
					EXTERNAL_NET += connection + ",";
			}
			EXTERNAL_NET += "]";
			LOG.i(this, "Launching snort");
			String snortConfiguration = FjoSpidie.configuration.getSnortConfig();
			pb = new ProcessBuilder("snort", "-c", snortConfiguration, "-A", "console", "-q", "-N", "-r",
					FjoSpidie.pcapPath, "-S", "EXTERNAL_NET=" + EXTERNAL_NET);

			analisysProgess = pb.start();
			BufferedReader input = new BufferedReader(new InputStreamReader(analisysProgess.getInputStream()));
			StreamGobbler err = new StreamGobbler(analisysProgess.getErrorStream(), "ERROR ");
			err.start();
			String line = "";
			while ((line = input.readLine()) != null) {
				SnortAlert alert = new SnortAlert(line);
				alerts.add(alert);
				Report.addAlert(alert);

				LOG.d(alert.getAlarm_text() + " " + alert.getClassification());

			}
			int exitVal = analisysProgess.waitFor();
			LOG.i(this, "Snort stopped " + exitVal);
			LOG.i(this, "SQLifying alerts");

		} catch (Exception e) {
			e.printStackTrace();
		}

	}
}

File: fjospidie/engine/SnortEngine.py
import threading
import logging
import subprocess
from snort.SnortAlert import SnortAlert
class SnortEngine(threading.Thread):
    def __init__(self, report, connections, snort_config, pcap_path):
        threading.Thread.__init__(self)
        self.connections = connections
        self.report = report
        self.snort_config = snort_config
        self.pcap_path = pcap_path
        self.alerts = []

    def run(self):
        logging.info("Starting SnortEngine")

        external_net = "["
        for connection in self.connections:
            external_net += connection + ','
        external_net += "]"

        snort_command = ["snort", "-c", self.snort_config, "-A", "console", "-q", "-N", "-r", self.pcap_path, "-S", "EXTERNAL_NET=" + external_net]
        logging.info("Running snort: " + " ".join(snort_command))
        snort = subprocess.Popen(snort_command, stdout=subprocess.PIPE)
        while True:
            line = snort.stdout.readline()
            if line != '':
                alert = SnortAlert(line)
                self.alerts.append(alert)
            else:
                break
        self.report.add_alerts( self.alerts)
        logging.info("Stopping SnortEngine")
